File: commands/general/commandstats.ts
import { ChatInputCommandInteraction, EmbedBuilder, SlashCommandBuilder, Snowflake } from 'discord.js';
import Bot from '../../bot/bot';
import SlashCommand from '../../bot/slashcommand';
import UtilityFunctions from '../../utils/utilities';

export default class CommandStats extends SlashCommand {
  constructor(client: Bot) {
    const data = new SlashCommandBuilder()
      .setName('commandstats')
      .setDescription('View usage statistics on a command')
      .setDMPermission(false)
      //.setAutocomplete(true)
      .addStringOption(option => option
        .setName('command-name')
        .setDescription('The command to view stats for. Category name included')
        .setRequired(true)
      );
    super(data as SlashCommandBuilder, client);
  }

  async execute(interaction: ChatInputCommandInteraction): Promise<void> {
    const commandName = interaction.options.getString('command-name')!.split(/\s+/).join(' ');

    let commandUsages: Array<[Snowflake, number]>;
    
    try {
      commandUsages = await this.client.database.getCommandUsage(interaction.guildId!, commandName);
    }
    catch (error) {
      const embed = new EmbedBuilder()
        .setColor(this.client.config.colors.RED)
        .setTitle('Command stats not found')
        .setDescription('No stats for this command found in this server.\nDid you spell everything correctly?');
      interaction.reply({ embeds: [embed], ephemeral: true });
      return;
    }

    const embeds = UtilityFunctions.chunk(commandUsages, 20).map(usageList => {
      const embed = new EmbedBuilder()
        .setColor(this.client.config.colors.BLUE)
        .setTitle(`Command stats: ${commandName}`)
        .addFields({
          name: 'Top usage',
          value: usageList.map(userInfo => `${userInfo[1]} - <@${userInfo[0]}>`).join('\n')
        });
      return embed;
    });

    if (embeds.length > 1) {
      this.sendMenu(interaction, embeds);
    }
    else {
      interaction.reply({ embeds: [embeds[0]] });
    }
  }
}

File: src/commands/general/commandstats.rs
use regex::Regex;
use serenity::all::{CommandInteraction, CommandOptionType};
use serenity::async_trait;
use serenity::builder::{CreateCommand, CreateCommandOption, CreateEmbed};
use serenity::client::Context;

use crate::color::Colors;
use crate::commands::SlashCommand;
use crate::database::Database;
use crate::interaction::pager::InteractiveMenu;
use crate::interaction::{BetterResponse, InteractionCustomGet};
use crate::Result;

#[derive(Default)]
pub struct CommandStats;

#[async_trait]
impl SlashCommand for CommandStats {
  fn register(&self) -> CreateCommand {
    CreateCommand::new("commandstats")
      .description("View usage statistics on a command")
      .dm_permission(false)
      .add_option(
        CreateCommandOption::new(
          CommandOptionType::String,
          "command-name",
          "The full name of the command to view stats for"
        )
        .required(true)
      )
  }

  async fn execute(&self, ctx: &Context, interaction: &CommandInteraction, db: &Database) -> Result<()> {
    let re = Regex::new(r"\s+").unwrap();
    let input = interaction.get_string("command-name").unwrap();
    let command_name = re
      .split(input.as_str())
      .filter(|m| !m.is_empty())
      .collect::<Vec<_>>()
      .join(" "); // Sanitize

    match db.get_command_usage(command_name.clone(), interaction.guild_id).await {
      Ok(result) => {
        let embeds = result
          .chunks(20)
          .map(|stats| {
            let names = stats.iter().map(|(id, _)| format!("<@{id}>")).collect::<Vec<_>>();
            let usage = stats.iter().map(|(_, count)| count.to_string()).collect::<Vec<_>>();

            CreateEmbed::new()
              .color(Colors::Blue)
              .title("Top Usage")
              .description(format!("Command: `/{command_name}`"))
              .fields([("Name", names.join("\n"), true), ("Usage", usage.join("\n"), true)])
          })
          .collect::<Vec<_>>();

        if embeds.len() > 1 {
          interaction.send_menu(ctx, embeds).await?;
        } else {
          interaction.reply_embed(ctx, embeds[0].clone()).await?;
        }
      }

      Err(_) => {
        let embed = CreateEmbed::new()
          .color(Colors::Red)
          .title("Command stats not found")
          .description("No stats for that command found in this server.\nDid you spell & format everything correctly?");

        interaction.reply_embed_ephemeral(ctx, embed).await?;
      }
    }

    Ok(())
  }
}
